scripts/ioc_deploy.py

```diff
@@ -301,6 +426,143 @@ def make_in(deploy_dir: str, dry_run: bool) -> int:
         return subprocess.run(["make"], cwd=deploy_dir).returncode
 
 
+def set_permissions(deploy_dir: str, allow_write: bool, dry_run: bool) -> int:
+    """
+    Apply or remove write protection from a deploy repo.
+
+    You may or may not have permissions to do this to releases created by other users.
+
+    allow_write=True involves adding "w" permissions
+    for the owner and for the group. "w" permissions will never be added for other users.
+    We will also add write permissions to the top-level directory.
+
+    allow_write=False involves removing the "w" permissions
+    from all files, and from directories that contain untracked files.
```

Comment:
Maybe I'm just slow here, but what's the difference between the files affected when `allow_write=True` and `allow_write=False`?  
* `allow_write=True`  -->  recursively applies to everything inside `deploy_dir` except for ".git"
* `allow_write=False` --> everything tracked by git and their subdirectories recursively, excepting ".git", but then also things not tracked by git?

Is there something missing from the union of "things tracked by git" and "things not tracked by git"?

Reply:
Maybe I haven't internalized what the directory structure here looks like

Reply:
This is a good question that might prompt some edits

I think it's normal to expect that the set of files affected by these two is the same- maybe I should edit so that it is.

The perspective here is something like "there are some directory objects that should never be write-protected because the group wants to be ably to casually recompile on a new arch", but there is no such constraint in the other direction.

- `allow_write=True` -> no constraints, (re-)implemented separately in the simplest way without constraint logic
- `allow_write=False` -> please allow my to be able to log into the rocky9 build host and create e.g. `bin/rhel9_x86_64` even if I've already built/protected once e.g. `bin/rhel7_x86_64`

When implementing the `False` variant with the above goal, I decided that a simple way to do this is to write protect directories that contain files generated from `make`, e.g. those not tracked by git.

Cons of this approach:

- Confusing code
- git safedir thing will pop up when you try to write protect someone else's release

Alternates I considered that maybe you think are better:

- Make it simpler: full write permissions or no write permissions for the entire tree, require using the script to unlock for a new make
- Go for the minimal number of allowed directories: look for folders with OS-names in them, allow only those directories to stay unprotected

I ignored .git because it seemed weird to be messing with it but maybe it's better to mess with it for simplicity's sake

Open to all thoughts and suggestions here

Reply:
Easy agree: I think ignoring .git makes perfect sense

Ok so the difference is that when we set `allow_write=False`, we do not write protect the build artifact directories generated by `make build`

It looks like the code is gathering `build_dir_paths`, then also write-protecting them.  Do we actually want to make these writable?  (are these the `bin/rhel_xx_xx` directories?)

Reply:
The `make` generates files and folders such as:
```
bin/rhel7_x86_64/adsIoc
```
Which is the main executable that needs to continue to exist in this example
The intention is to write protect `bin/rhel7_x86_64` without write protecting `bin`:
```
── [drwxrwxr-x]  bin
│   └── [dr-xr-xr-x]  rhel7-x86_64
│       └── [-r-xr-xr-x]  adsIoc
```
So that we can't accidentally remove the folder and delete `adsIoc`, but we can create a parallel folder in `bin` for a new arch.